### PR TITLE
Make search work in Frontend

### LINKED ISSUE #306: Make "Search" work
Splitted from #42

### PR DESCRIPTION
**Pull request by bobdenotter**
Fixes #306 

todo:

 - [x] Update search template
 - [x] Make search controller work 
 - [x] Implement naive search
 - [x] Add tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -69,10 +69,9 @@ e2e-wip:
 	cd tests/e2e && npm run kakunin -- --tags @wip && cd ../..
 
 e2e-install:
-	cd tests/e2e
-	npm install
-	node ./node_modules/protractor/bin/webdriver-manager update --gecko=false
-	cd step_definitions
+	cd tests/e2e && npm install
+	node ./tests/e2e/node_modules/protractor/bin/webdriver-manager update --gecko=false
+	cd tests/e2e/step_definitions
```
Comment (JarJak) [inline Makefile]: what was the problem here? do symlinks below work?

File: public/theme/skeleton/search.twig [hunk 2]
```diff
@@ -18,7 +18,7 @@
         {% setcontent block = "block/search-teaser" %}
```
Comment (JarJak) [inline public/theme/skeleton/search.twig]: rename block to searchTeaser

File: public/theme/skeleton/search.twig [hunk 3]
```diff
@@ -18,7 +18,7 @@
...
 
         {# check if we have 'content'. If so, we know we have have a teaser to display. #}
-        {% if block is defined and block.hasField('content') %}
+        {% if block and block.content %}
```
Comment (JarJak) [inline public/theme/skeleton/search.twig]: rename content to something more meaningful

File: public/theme/skeleton/search.twig [hunk 4]
```diff
@@ -47,16 +47,14 @@
...
-                    {{ __('general.phrase.no-result-for-search-term', { '%SEARCHTERM%': search|escape }) }}
+                    {{ __('general.phrase.no-search-results-for', { '%search%': search|escape }) }}
                 </p>
 
-            </article>
         {% endfor %}
 
         {# If there are more records than will fit on one page, the pager is shown. #}
-        {{ pager(template = 'partials/_sub_pager.twig') }}
+        {{ pager(records, template = 'partials/_sub_pager.twig') }}
```
Comment (JarJak) [inline public/theme/skeleton/search.twig]: this template is default so can be omitted here

File: src/Repository/ContentRepository.php [hunk 5]
```diff
@@ -67,6 +70,41 @@ public function findLatest(?ContentType $contentType = null, int $amount = 6): ?
         return $qb->getQuery()->getResult();
     }
 
+    public function findNaiveSearch(int $page = 1, string $search = '', int $amountPerPage = 6, bool $onlyPublished = true)
+    {
+        if (empty($search)) {
+            return null;
```
Comment (JarJak) [inline src/Repository/ContentRepository.php]: move input validation to controller

File: src/Repository/ContentRepository.php [hunk 6]
```diff
@@ -67,6 +70,41 @@ public function findLatest(?ContentType $contentType = null, int $amount = 6): ?
         return $qb->getQuery()->getResult();
     }
 
+    public function findNaiveSearch(int $page = 1, string $search = '', int $amountPerPage = 6, bool $onlyPublished = true)
```
Comment (JarJak) [inline src/Repository/ContentRepository.php]: rename: searchNaive

File: src/Repository/ContentRepository.php [hunk 7]
```diff
@@ -67,6 +70,41 @@ public function findLatest(?ContentType $contentType = null, int $amount = 6): ?
...
+        $qb = $this->getQueryBuilder()
+            ->select('partial content.{id}');
+
+        $qb->addSelect('f')
+            ->leftJoin('content.fields', 'f')
+            ->andWhere($qb->expr()->like('f.value', ':search'))
+            ->setParameter('search', '%' . $search . '%');
+
+        // These are the ID's of content we need.
+        $ids = new Collection($qb->getQuery()->getArrayResult());
```
Comment (JarJak) [inline src/Repository/ContentRepository.php]: $ids = array_column($qb->getQuery()->getScalarResult(), "id")

File: src/Twig/RecordExtension.php [hunk 8]
```diff
@@ -75,8 +75,12 @@ public function dummy_with_env(Environment $env, $input = null)
         return $input;
     }
 
-    public function pager(Environment $env, Pagerfanta $records, string $template = '_sub_pager.twig', string $class = 'pagination', string $theme = 'default', int $surround = 3)
+    public function pager(Environment $env, ?Pagerfanta $records, string $template = '_sub_pager.twig', string $class = 'pagination', string $theme = 'default', int $surround = 3)
```
Comment (JarJak) [inline src/Twig/RecordExtension.php]: rename $env to $twig (everywhere in this class)

Comment (JarJak) [inline src/Twig/RecordExtension.php]: how can $records be null instead of empty array / empty pagerfanta object?

File: src/Twig/RecordExtension.php [hunk 9]
```diff
@@ -75,8 +75,12 @@ public function dummy_with_env(Environment $env, $input = null)
...
     {
+        if (! $records) {
```
Comment (JarJak) [inline src/Twig/RecordExtension.php]: if (count($records) === 0)

File: tests/e2e/features/display_search.feature [hunk 10]
```diff
@@ -0,0 +1,18 @@
+Feature: Display search functionality
+    @wip
+    Scenario: As a user I want to display search results
+        When I visit the "homepage" page
+
+        Then I scroll to the "search_input" element
+        When I fill the "search_input" field with "consequatur"
+        And I click the "search_button" element
+
```
Comment (JarJak) [inline tests/e2e/features/display_search.feature]: add "search" page

File: tests/e2e/pages/homepage.js [hunk 11]
```diff
@@ -8,6 +8,13 @@ class HomepagePage extends BasePage {
 
     this.title = $('header a');
     this.recent_pages_list = $('aside ul');
+
+    // For testing "search"
```
Comment (JarJak) [inline tests/e2e/pages/homepage.js]: move to "search" page

File: Makefile (hunk 1 above)
Comment (bobdenotter) [reply Makefile]: I'm not sure.. it was this: 

```
cd tests/e2e
npm install
```

Which resulted in `make` running `npm install` in the project root, instead of in the folder it `cd`'ed to in the previous command.

Comment (bobdenotter) [reply Makefile]: The symlinks do work, yes.

File: public/theme/skeleton/search.twig (hunk 3 above)
Comment (bobdenotter) [reply public/theme/skeleton/search.twig]: The "Blocks" ContentType has a `content` field. It's generic on purpose :-)

File: public/theme/skeleton/search.twig (hunk 4 above)
Comment (bobdenotter) [reply public/theme/skeleton/search.twig]: In the past i've experienced that it's helpful for people who want to tweak it, to have this explicit. We've had a _lot_ of people asking "How do I change the pager template?"

File: src/Twig/RecordExtension.php (hunk 8 above)
Comment (bobdenotter) [reply src/Twig/RecordExtension.php]: I'll make a separate PR for that. See #368

Comment (bobdenotter) [reply src/Twig/RecordExtension.php]: People invoking `{{ pager(records) }}` when records is null. For instance on an empty search page.

File: tests/e2e/features/display_search.feature (hunk 10 above)
Comment (bobdenotter) [reply tests/e2e/features/display_search.feature]: What do you mean?

File: src/Twig/RecordExtension.php (hunk 8 above)
Comment (JarJak) [reply src/Twig/RecordExtension.php]: on empty search page, `records` should be an empty iterable, not null

File: tests/e2e/features/display_search.feature (hunk 10 above)
Comment (JarJak) [reply tests/e2e/features/display_search.feature]: Add new PageObject and then check if you are on right page

File: tests/e2e/pages/homepage.js (hunk 11 above)
Comment (bobdenotter) [reply tests/e2e/pages/homepage.js]: You'll have to show me how.. I tried making a search.js for it, but the "feature" starts on "homepage" -> `When I visit the "homepage" page`. From there it navigates to the search page, but the orignal 'page' still applies to it, as far as the javascript is concerned.

Comment (JarJak) [reply tests/e2e/pages/homepage.js]: `Then the "search" page is displayed` and it should switch current page object

File: tests/e2e/features/display_search.feature (hunk 10 above)
Comment (bobdenotter) [reply tests/e2e/features/display_search.feature]: How? :-)

File: src/Repository/ContentRepository.php (hunk 7 above)
Comment (bobdenotter) [reply src/Repository/ContentRepository.php]: Yes, nice!

File: src/Repository/ContentRepository.php (hunk 6 above)
Comment (bobdenotter) [reply src/Repository/ContentRepository.php]: done

File: tests/e2e/pages/homepage.js (hunk 11 above)
Comment (bobdenotter) [reply tests/e2e/pages/homepage.js]: Done!

File: src/Controller/Frontend/SearchController.php [hunk 12]
```diff
@@ -37,12 +37,23 @@ public function __construct(Config $config, Environment $twig, TemplateChooser $
...
         $page = (int) $request->query->get('page', 1);
+        $search = $request->get('search', '');
+        $amountPerPage = $this->config->get('general/listing_records');
 
         // @todo implement actual Search Engine
-        $records = $contentRepository->findForListing($page);
+        if (! empty($search)) {
+            $records = $contentRepository->searchNaive($search, $page, $amountPerPage);
+        } else {
+            $records = null;
```
Comment (JarJak) [inline src/Controller/Frontend/SearchController.php]: it should be empty array, otherwise pagerfanta etc. will break ;)

Comment (JarJak) [reply src/Controller/Frontend/SearchController.php]: also add a test what will happen if query is empty

Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: I _think_ `null` is correct, and it should be `?Pagerfanta $records`. I've added a test to illustrate. :-)

Schermafbeelding 2019-03-27 om 16 08 14

Comment (JarJak) [reply src/Controller/Frontend/SearchController.php]: yes, it gives an exception because this line returns null instead of empty array

File: src/Controller/Frontend/SearchController.php [hunk 13]
```diff
@@ -37,12 +37,23 @@ public function __construct(Config $config, Environment $twig, TemplateChooser $
...
+        }
+
+        $context = [
+            'search' => $search,
```
Comment (JarJak) [inline src/Controller/Frontend/SearchController.php]: rename search to searchQuery or searchedText

File: src/Controller/Frontend/SearchController.php (hunk 12 above)
Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: No, it _breaks_ as soon as that line returns `[]`. If the method accepts only `Pagerfanta $records`, it does break if you're trying to pass in `[]`.. 

On the other hand, `?Pagerfanta $records` allows you to pass in `null` and handle it properly.

File: src/Controller/Frontend/SearchController.php (hunk 13 above)
Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: I'd prefer not. The parameter (in `POST` or `GET`) is also `search`, and it has always been this way historically.

File: src/Repository/ContentRepository.php [hunk 14]
```diff
@@ -67,6 +69,37 @@ public function findLatest(?ContentType $contentType = null, int $amount = 6): ?
         return $qb->getQuery()->getResult();
     }
 
+    public function searchNaive(string $search, int $page, int $amountPerPage, bool $onlyPublished = true)
```
Comment (JarJak) [inline src/Repository/ContentRepository.php]: add return type

File: src/Controller/Frontend/SearchController.php (hunk 13 above)
Comment (JarJak) [reply src/Controller/Frontend/SearchController.php]: `historically` ;) Let's move to the future!
We can break anything from past Bolt versions, can't we?
Also some badly named things?

File: src/Controller/Frontend/SearchController.php (hunk 12 above)
Comment (JarJak) [reply src/Controller/Frontend/SearchController.php]: Right, so let's return empty Pagerfanta then, to keep context interface clean.

Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: Ok, sounds good. Done.

File: src/Controller/Frontend/SearchController.php (hunk 13 above)
Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: No, we don't have to keep it, but I think it's a properly named thing, in this case. :-)

File: src/Repository/ContentRepository.php (hunk 14 above)
Comment (bobdenotter) [reply src/Repository/ContentRepository.php]: Done

File: src/Controller/Frontend/SearchController.php (hunk 13 above)
Comment (stryxx) [reply src/Controller/Frontend/SearchController.php]: Gentlemen, please :D Variable name "searchValue" exactly says that it is a value which is searching :) "search" means "activity" we do.

Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: Alright, alright.. :-)

I've renamed it to `searchTerm` and made the same change in the default templates. I couldn't help myself from keeping the old one backwards compatible, though.

File: Makefile [hunk 15]
```diff
@@ -69,18 +69,8 @@ e2e-wip:
 	cd tests/e2e && npm run kakunin -- --tags @wip && cd ../..
 
 e2e-install:
-	cd tests/e2e
-	npm install
-	node ./node_modules/protractor/bin/webdriver-manager update --gecko=false
-	cd step_definitions
```
Comment (JarJak) [inline Makefile]: why removed?

File: public/theme/skeleton/search.twig [hunk 16]
```diff
@@ -47,16 +47,20 @@
             {% endif %}
 
         {% else %}
-            <article>
 
-                <p>
-                    {{ __('general.phrase.no-result-for-search-term', { '%SEARCHTERM%': search|escape }) }}
+            {% if search is not empty %}
```
Comment (JarJak) [inline public/theme/skeleton/search.twig]: searchTerm

File: src/Controller/Frontend/SearchController.php [hunk 17]
```diff
@@ -37,13 +39,24 @@ public function __construct(Config $config, Environment $twig, TemplateChooser $
...
-        $records = $contentRepository->findForListing($page, $amountPerPage);
+        if (! empty($searchTerm)) {
+            $records = $contentRepository->searchNaive($searchTerm, $page, $amountPerPage);
+        } else {
+            $records = new Pagerfanta(new ArrayAdapter([]));
+        }
+
+        $context = [
+            'searchTerm' => $searchTerm,
+            'search' => $searchTerm, // Keep 'search' for Backwards Compatibility
```
Comment (JarJak) [inline src/Controller/Frontend/SearchController.php]: what breaks if you remove this line?

File: Makefile (hunk 15 above)
Comment (bobdenotter) [reply Makefile]: Because the files are in git, these symlinks weren't created properly, but we didn't notice yet, because they're already there: https://github.com/bolt/four/tree/master/tests/e2e/step_definitions

File: public/theme/skeleton/search.twig (hunk 16 above)
Comment (bobdenotter) [reply public/theme/skeleton/search.twig]: Done, thanks.

File: src/Controller/Frontend/SearchController.php (hunk 17 above)
Comment (bobdenotter) [reply src/Controller/Frontend/SearchController.php]: Nothing in the default template. It'll break for people porting a Bolt 3 template, and/or people who are used to using `search`. It's a small effort to support it, and the impact is very low, but it will save some people some headaches.

**Comment by bobdenotter** [comment]
Thanks!
